tests/main/apparmor-batch-reload/task.yaml

```diff
@@ -0,0 +1,65 @@
+summary: Ensure that batch-reloading of AppArmor profiles works, including fallback
+
+systems: [ubuntu-1*]
+
+environment:
+    VARIANT/changed: changed
+    VARIANT/unchanged: unchanged
+    AA_CACHE: /var/cache/apparmor/
+    AA_PROFILES: /var/lib/snapd/apparmor/profiles/
+
+prepare: |
+    snap install test-snapd-content-plug test-snapd-tools
+    snap install --edge test-snapd-hello-multi-arch
+    cp /sbin/apparmor_parser /sbin/apparmor_parser.real
+    rm -f /tmp/apparmor-parser-calls.debug
+
+restore: |
+    mv /sbin/apparmor_parser.real /sbin/apparmor_parser
+    rm -f /tmp/apparmor-parser-calls.debug
+
+debug: |
+    cat /tmp/apparmor-parser-calls.debug || true
+
+execute: |
+    systemctl stop snapd.{service,socket}
+
+    echo "Update system key"
+    printf '{"version":1}' > /var/lib/snapd/system-key
+
+    # this makes is easier to analyse the sequence of events shall something fail
+    echo "system key update" >> /tmp/apparmor-parser-calls.debug
+
+    echo "Replace apparmor parser with a broken one"
+    cp bin/apparmor_parser.fake /sbin/apparmor_parser
+
+    # remove all apparmor cached profiles so we can check they are recreated
+    rm -f $AA_CACHE/snap*
+
+    # "changed" variant requires source profiles on the disk to look different than the
+    # ones generated on snapd restart, simulate this by appending a comment to them.
+    if [ "$VARIANT" = "changed" ]; then
+        for profile in $AA_PROFILES/snap*; do
+            echo "# forced profile change" >> $profile
+        done
+    fi
+    systemctl start snapd.{socket,service}
+
+    echo "Checking that apparmor profile cache has been populated"
+    # note, the list is not exhaustive but is enough for the test and the journal check below
+    for p in snap-update-ns.test-snapd-hello-multi-arch snap.test-snapd-hello-multi-arch.hello-i386 snap-update-ns.test-snapd-tools snap.test-snapd-tools.block snap.test-snapd-tools.cat; do
+        retry-tool -n 5 test -f $AA_CACHE/$p
```

Comment:
fyi, 5 seconds is not necessarily a long time on, for example, an ARM device. Since this is a subset of profiles, I worry that this approach may be racy since apparmor_parser will parse these in the order given and it is possible there could be a few profiles prior to one of the ones in the list that will take longer and push this to fail. I'm not sure how you want to reorganize this, but you could wait on ps output to no longer have any apparmor_parsers and then perform your checks (though that doesn't work for the fallback case where the parsers are being called one after the other). You could calculate the total time to wait on that as something terrible like (10*number of profiles in AA_PROFILES).

Reply:
Thank you, that's a good observation, I changed it to retry a single test -f check over multiple profiles at once, with a more generous number of retries.

Reply:
I actually re-worked it slightly to use `find` to discover profiles in AA cache, and then diff them with expected set. This should make it happy with "cache as a forest" in newer apparmor versions (thanks for pointing this AA change out!).